lib/pure/collections/sequtils.nim

```diff
@@ -774,48 +781,86 @@ template anyIt*(s, pred: untyped): bool =
 
-template toSeq2(iter: iterator): untyped =
-  # overload for iterator
-  evalOnceAs(iter2, iter(), false)
-  when compiles(iter2.len):
-    var i = 0
-    var result = newSeq[typeof(iter2)](iter2.len)
-    for x in iter2:
-      result[i] = x
-      inc i
-    result
+template toSeqBuiltin[T](arg: openArray[T]): seq[T] = @arg
+
+template toSeqBuiltin(arg: string): seq[char] = @arg
+
+template toSeqBuiltin[T](arg: seq[T]): seq[T] = arg
+
+proc toSeqBuiltin[T](arg: set[T]): seq[T] {.inline.} =
+  for it in arg:
+    result.add it
+
+proc toSeqBuiltin(a: cstring): seq[char] {.inline.} =
+  let len = a.len
+  result.setLen len
+  var i = 0
+  while i < len:
+    result[i] = a[i]
+    inc i
+
+proc toSeqBuiltin[E: enum](t: typedesc[E]): seq[E] =
+  for it in t:
+    result.add it
+
+macro toSeqBuiltin(arg: iterator): untyped =
+  let typ = arg.getTypeInst
+  typ.expectKind nnkProcTy # only for closure iterators
+  let formalParams = typ[0]
+  formalParams.expectKind nnkFormalParams
+  formalParams.expectMinLen 1
+  let resultTyp = nnkBracketExpr.newTree(bindSym"seq", formalParams[0])
+  let tmpSym = genSym(nskVar)
+  result = nnkStmtListExpr.newTree
+  result.add nnkVarSection.newTree(nnkIdentDefs.newTree(tmpSym, resultTyp, newNimNode(nnkEmpty)))
+  result.add quote do:
+    for it in `arg`:
+      `tmpSym`.add it
+  result.add tmpSym
+
+template toSeqBuiltin[T: not proc](arg: T): untyped =
+  # fallback for ``IntSet`` and ``HashSet``. Explicit overload would
+  # be better, but this would require an additional import here.
+  var tmp: seq[typeof(items(arg), typeOfIter)]
+  for it in arg:
+    tmp.add it
+  tmp
+
+macro toSeqImpl(arg: typed): untyped =
+  let iteratorCall = arg[1]
+  let iteratorSym = iteratorCall[0]
+  if iteratorCall.len == 2 and iteratorSym.eqIdent("items") and iteratorSym.owner.owner.eqIdent("stdlib"):
+    let iteratorArg = iteratorCall[1]
+    result = newCall(bindSym"toSeqBuiltin", iteratorArg)
   else:
-    type OutType = typeof(iter2())
-    var result: seq[OutType] = @[]
-    when compiles(iter2()):
-      evalOnceAs(iter4, iter, false)
-      let iter3 = iter4()
-      for x in iter3():
-        result.add(x)
-    else:
-      for x in iter2():
-        result.add(x)
-    result
+    var typ = iteratorCall.getTypeInst()
+    # `lent` is broken feature. and this special handling to get rid of lent is why it is broken.
+    if typ.kind == nnkBracketExpr and typ.len == 2 and typ[0].eqIdent("lent"):
+      # TODO this is not guaranteed to be the acutal system `lent` symbol.
+      typ = typ[1]
+
+    result = quote do:
+      var tmp: seq[`typ`]
+      for it in `iteratorCall`:
+        tmp.add it
+      tmp
+
+proc genCallToSeqImpl(arg: NimNode): NimNode =
+  newCall(bindSym"toSeqImpl",
+    nnkForStmt.newTree(
+      ident"x", arg, newStmtList()))
+
+macro toSeqImplSingleSym(arg: typed): untyped =
+  if arg.kind != nnkSym:
+    error "identifier not unique", arg # arg is symChoice
```

Comment:
I think it's better to generate a call for it and have the compiler ambiguous call handling trigger instead, which would give a better error message.

Reply:
Sorry I don't know what you mean, or if it is possible. Again it is really long ago since I wrote it, but knowing myself, I certainly would have done it that way, if I would have seen it feasible.

Reply:
I mean to just `genCallToSeqImpl(newCall(arg))` it, which passes the burden downstream for the compiler to handle.